Support MELPA package empv.el

Boruch-Baum opened the pull request:
+ Check which of TWO Emacs mpv interfaces are loaded and active, and
  uses that one to get data for creating org-links and notes.

+ Report errors for 'no backend package loaded' and 'no data received
  from the backend'

Boruch-Baum committed:
Support MELPA package empv.el

+ Check which of TWO Emacs mpv interfaces are loaded and active, and
  uses that one to get data for creating org-links and notes.

+ Report errors for 'no backend package loaded' and 'no data received
  from the backend'

bpanthi977 commented:
@Boruch-Baum I have refactored the code and added support for other functionalities too. Please verify with empv.

Boruch-Baum commented:
;tldr Qualified success. Unrelated bugs and quirks uncovered.

+ The first hurdle to proper testing with empv only is the elisp header
  line "Package-Requires:" which includes mpv. I posted a question to
  MELPA support and it turns out that the header doesn't support
  alternatives, so the only way I succeeded to begin this test was to
  remove the requirement in that line for mpv, remove package
  org-empv-notes (MELPA version), remove package mpv, restart Emacs
  (unloading the features tuned out to be insufficient for some reason),
  manually evaluate the buffer, and manually load the feature. I can
  then verify that package mpv is not loaded or present.

  + IMO it's desirable to remove package mpv from that header, and rely
    on the run-time error messages to guide the user to select the
    backend of the user's choice.

+ The next observation regards deprecated functions. Emacs has been
  nudging people to stop using 'find' (since v27) in favor 'cl-find' and
  'flet' (since v24) in favor of 'cl-flet'.

+ The code successfully allowed me to create notes and links using empv.

+ The following functions succeed:
    org-mpv-notes-speed-{up,down}
    org-mpv-notes-toggle-fullscreen
    org-mpv-notes-save-screenshot

  + IMO functions org-mpv-notes-speed-{up,down} are examples of features
    that could just as well be left to the backend interfaces (mpv.el,
    empv.el).

    On that subject, I couldn't figure out a way to start a NEW video
    using org-mpv-notes, so I made function org-mpv-notes-open
    (interactive "f"), which works with the empv.el backend and couldbe
    useful if you want to go the route of wrapping backend features.
    Currently, doing so fails silently on files not supported by mpv.

+ The following functions have not been modified (and I don't understand
  what they do anyway):
    org-mpv-notes-seek-halve-step
    org-mpv-notes-seek-double-step

+ I'll create issues for the unrelated bugs and quirks uncovered.

[1] https://github.com/melpa/melpa/issues/8830#issuecomment-1823642034

--
hkp://keys.gnupg.net
CA45 09B5 5351 7C11 A9D1  7286 0036 9E45 1595 8BC0

Boruch-Baum commented:
Also:

+ Remove the TODO comment in function org-mpv-notes-open ?

--
hkp://keys.gnupg.net
CA45 09B5 5351 7C11 A9D1  7286 0036 9E45 1595 8BC0

Boruch-Baum committed:
Resolve Bibek's force-push commit 91bfc2a

+ Isolate Bibek's work as a separate commit

+ Remove hunks that were committed on other branches, for other
  features

Boruch-Baum committed:
Remove requirement for MELPA package 'mpv'

+ This is strongly suggested because having more than one mpv-related
  back-end (eg. mpv.el and empv.el) creates a confusion point for the
  user with the possiblity of attempting to mix and match between the
  two backends.

+ Emacs currently does not support an 'alternatives' feature for the
  'Package-Requires' header (ie. logical 'or').

+ The disadvantage of this commit is that users will need to know to
  have installed an mpv back-end before using this package. Explicit
  error messages to that effect, along with advice, already exist in the
  code.

+ The advantage of this commit, well, paragraph one, above.

Boruch-Baum committed:
Resolve warnings of deprecated functions

+ Replace 'find' and 'flet' with 'cl-find' and 'cl-flet'

bpanthi977 commented:
> On that subject, I couldn't figure out a way to start a NEW video
>    using org-mpv-notes, so I made function org-mpv-notes-open
>    (interactive "f"), which works with the empv.el backend and couldbe
>    useful if you want to go the route of wrapping backend features.
>    Currently, doing so fails silently on files not supported by mpv.

The way I do this is create an org link to the video ( [[mpv:/path/to/video]] ), then click on it. As you have demonstrated I should improve this. Your interactive function would be a good addition.

> IMO functions org-mpv-notes-speed-{up,down} are examples of features
>  that could just as well be left to the backend interfaces (mpv.el,
>  empv.el).

You have a good point. I think we should remove these functions from org-mpv-notes package. Users of the pacakage can just use the primitive provided by mpv or empv. Though it would be nice to move this to README.md showing a way to configure the package. My config is at https://github.com/bpanthi977/.emacs.d/blob/master/modules/org-config.el#L1471-L1497

> + The following functions have not been modified (and I don't understand
>  what they do anyway):
>    org-mpv-notes-seek-halve-step
>    org-mpv-notes-seek-double-step

These can also be removed. 

> Also: + Remove the TODO comment in function org-mpv-notes-open ?

Yes.

Boruch-Baum committed:
Remove TODO comment line

Boruch-Baum committed:
Seek to proper position on opening a new file

+ Prior behavior was to play from the start even when the timestamp
  indicated otherwise.

Boruch-Baum committed:
Add non-fatal requires for backends

+ This won't force an install of unwanted backend packages, and will
  guarantee that a package that is installed locally is loaded.

Boruch-Baum committed:
Debug commit: empv-support, seeking to a position

+ This commit is the result of a debugging session that uncovered some
  issues that are all arguably empv support

bpanthi977 commented:
Thank you for the PR. 

Most of the things in the PR had already been merged a while ago. I had cherry picked some of your commits. 

Today, I merged some code #20, #21  that fix some remaining issues. 

I am closing the PR. If you find that I missed something, feel free to leave a message. I'll pull your changes.
